File: medium/0253_meeting_rooms_ii/solution.py
import heapq
from typing import List

from utils.python.interval import Interval


class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        intervals.sort(key=lambda interval: interval.start)

        heap = []
        max_rooms = 0

        for interval in intervals:
            if heap and interval.start >= heap[0]:
                heapq.heappop(heap)

            heapq.heappush(heap, interval.end)

            if len(heap) > max_rooms:
                max_rooms = len(heap)

        return max_rooms

File: medium/0253_meeting_rooms_ii/solution.cpp
#include "../../utils/cpp/interval.hpp"
#include <algorithm>
#include <vector>

class Solution {
public:
  int minMeetingRooms(std::vector<utils::Interval> &intervals) {
    std::sort(intervals.begin(), intervals.end(),
              [](const utils::Interval &a, const utils::Interval &b) {
                return a.start < b.start;
              });

    std::vector<int> heap;

    int maxRooms = 0;

    for (const utils::Interval &interval : intervals) {
      if (!heap.empty() && interval.start >= -heap[0]) {
        std::pop_heap(heap.begin(), heap.end());
        heap.pop_back();
      }

      heap.push_back(-interval.end);
      std::push_heap(heap.begin(), heap.end());

      maxRooms = std::max(maxRooms, (int)heap.size());
    }

    return maxRooms;
  }
};
